src/externs.rs

```diff
@@ -57,84 +37,37 @@ pub unsafe extern fn memcpy(dest: *mut u8, src: *const u8,
 ///
 /// This faster implementation works by copying bytes not one-by-one, but in
 /// groups of 8 bytes (or 4 bytes in the case of 32-bit architectures).
-#[cfg(target_pointer_width = "64")]
-#[no_mangle]
-pub unsafe extern fn memmove(dest: *mut u8, src: *const u8,
-                             n: usize) -> *mut u8 {
-    if src < dest as *const u8 {
-        let n_64: usize = n/8; // Number of 64-bit groups
-        let mut i: usize = n_64*8;
-
-        // Copy 8 bytes at a time
-        while i != 0 {
-            i -= 8;
-            *((dest as usize + i) as *mut u64) =
-                *((src as usize + i) as *const u64);
-        }
-
-        let mut i: usize = n;
-
-        // Copy 1 byte at a time
-        while i != n_64*8 {
-            i -= 1;
-            *((dest as usize + i) as *mut u8) =
-                *((src as usize + i) as *const u8);
-        }
-    } else {
-        let n_64: usize = n/8; // Number of 64-bit groups
-        let mut i: usize = 0;
-
-        // Copy 8 bytes at a time
-        while i < n_64 {
-            *((dest as usize + i) as *mut u64) =
-                *((src as usize + i) as *const u64);
-            i += 8;
-        }
-
-        // Copy 1 byte at a time
-        while i < n {
-            *((dest as usize + i) as *mut u8) =
-                *((src as usize + i) as *const u8);
-            i += 1;
-        }
-    }
-
-    dest
-}
-
-// 32-bit version of the function above
-#[cfg(target_pointer_width = "32")]
 #[no_mangle]
 pub unsafe extern fn memmove(dest: *mut u8, src: *const u8,
                              n: usize) -> *mut u8 {
     if src < dest as *const u8 {
-        let n_32: usize = n/4; // Number of 32-bit groups
-        let mut i: usize = n_32*4;
+        let n_usize: usize = n/WORD_SIZE; // Number of word sized groups
+        let mut i: usize = n_usize*WORD_SIZE;
 
-        // Copy 4 bytes at a time
+        // Copy `WORD_SIZE` bytes at a time
         while i != 0 {
-            i -= 4;
-            *((dest as usize + i) as *mut u32) =
-                *((src as usize + i) as *const u32);
+            i -= WORD_SIZE;
+            *((dest as usize + i) as *mut usize) =
+                *((src as usize + i) as *const usize);
         }
 
         let mut i: usize = n;
 
         // Copy 1 byte at a time
-        while i != n_32*4 {
+        while i != n_usize*WORD_SIZE {
             i -= 1;
             *((dest as usize + i) as *mut u8) =
                 *((src as usize + i) as *const u8);
         }
     } else {
-        let n_32: usize = n/4; // Number of 32-bit groups
+        let n_usize: usize = n/WORD_SIZE; // Number of word sized groups
         let mut i: usize = 0;
 
-        // Copy 4 bytes at a time
-        while i < n_32 {
-            *((dest as usize + i) as *mut u32) =
-                *((src as usize + i) as *const u32);
-            i += 4;
+        // Copy `WORD_SIZE` bytes at a time
+        while i < n_usize {
```

Comment:
I think this (and similar sites in the rest of the file) should be `i <= n - WORD_SIZE`.
This way, you are copying only a fraction of the length here, and copy a byte at a time for almost all of it.

Reply:
Good point. This would be a bug in the previous code as well.

Reply:
Oh, it would also need an extra special case for `n < WORD_SIZE`, since they are unsigned. Unsigned ints are evil.

Reply:
Edited my suggestion, the original one had potential overflow in it.